sql/parser/expr.go

```diff
@@ -241,72 +241,156 @@ func (NullVal) String() string {
 
 // QualifiedName is a base name and an optional indirection expression.
 type QualifiedName struct {
-	Base     Name
-	Indirect Indirection
+	Base       Name
+	Indirect   Indirection
+	tableName  bool
+	columnName bool
 }
 
 // StarExpr is a convenience variable that represents an unqualified "*".
 var StarExpr = &QualifiedName{Indirect: Indirection{unqualifiedStar}}
 
+// NormalizeTableName normalizes the qualified name to contain a database name
+// as prefix, returning an error if unable to do so or if the name is not a
+// valid table name (e.g. it contains an array indirection).
+func (n *QualifiedName) NormalizeTableName(database string) error {
```

Comment:
since these methods are all on the same type, i think they may be better named as `NormalizeAsFooName`